compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -6566,6 +6566,48 @@ boolean isNeverTypeOrStructureTypeWithARequiredNeverMember(BType type, Set<BType
         }
     }
 
+    public boolean isContainClientObjectOrFunctionType(BType type, boolean isCheckForClient) {
+        Set<BType> visitedTypeSet = new HashSet<>();
+        visitedTypeSet.add(type);
+        return isContainClientObjectOrFunctionType(type, visitedTypeSet, isCheckForClient);
+    }
+
+    boolean isContainClientObjectOrFunctionType(BType type, Set<BType> visitedTypeSet, boolean isCheckForClient) {
+        if (type == null) {
+            return false;
+        }
+        type = Types.getReferredType(type);
+        if (isCheckForClient && Symbols.isFlagOn(type.tsymbol.flags, Flags.CLIENT)) {
+            return true;
+        }
+        switch (type.tag) {
+            case TypeTags.INVOKABLE:
+                return !isCheckForClient;
+            case UNION:
+                for (BType memberType: ((BUnionType) type).getMemberTypes()) {
+                    if (!visitedTypeSet.add(memberType)) {
```

Comment:
What is the purpose of this `visitedTypeSet`?

Reply:
It store the previously visited datatypes. It helps to avoid visiting a previously visited datatype.
As an example, if we consider the a tuple of type `[typeA, typeB, typeA]`, this `visitedTypeSet` use to avoid checking the `typeA` again.